generator/template.go

```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
     }
   })
 }
+
+/**
+ * Flattens a deeply nested object
+ *
+ * @param  {Object} deeplyNestedObject
+ * @param  {String} path
+ * @return {Record<string, unknown>}
+ */
+function flattenObject<T extends Record<string, unknown>>(
+  deeplyNestedObject: T,
+  path: string = ""
+): T {
+  return Object.keys(deeplyNestedObject).reduce(
+    (acc: T, key: string): T => {
+      const value = deeplyNestedObject[key];
+
+      const newPath = Array.isArray(deeplyNestedObject)
+        ? path + "." + key
+        : [path, key].filter(Boolean).join(".");
+      const isObject = [
+        typeof value === "object",
+        value !== null,
+        !(value instanceof Date),
+        !(value instanceof RegExp),
+        !(Array.isArray(value) && value.length === 0)
+      ].every(Boolean);
+
+      return isObject
+        ? { ...acc, ...flattenObject(value as Record<string, unknown>, newPath) }
+        : { ...acc, [newPath]: value };
+    },
+    {} as T
+  );
+}
+
+/**
+ * Renders a deeply nested object into a string of URL search parameters
+ * by first flattening the object and then removing non-primitive array keys,
+ * non-primitive values and keys which are already present in the URL path.
+ * @param  {Object} deeplyNestedObject
+ * @param  {Array<string>} urlPathParams
+ * @return {String}
+ */
+export function renderURLSearchParams<T extends Record<string, unknown>>(
+  deeplyNestedObject: T,
+  urlPathParams: string[] = []
```

Comment:
urlPathParams is required here because `renderURLSearchParams` takes the whole request object as its input which also contains the key that may be present in the URL path as a parameter. So we need to ensure that fields in the request object that are present as URL path parameters are removed.

For example. Lets take the following RPC with endpoint:
```proto
message HTTPGetWithURLSearchParamsRequest {
  int32 a = 1;
  PostRequest post_req = 2;
  repeated int32 c = 3;
  ExternalMessage ext_msg = 4;
}

message HTTPGetWithURLSearchParamsResponse {
  int32 url_search_params_result = 1;
}

service CounterService {
  rpc HTTPGetWithURLSearchParams(HTTPGetWithURLSearchParamsRequest) returns (HTTPGetWithURLSearchParamsResponse) {
    option (google.api.http) = {
      get: "/api/query/{a}"
    };
  }
}
```

`renderURLSearchParams` takes the following whole request object as the input which contains the field `a` that is also present as the URL path parameter. 

```js
const req = {
  a: 10,
  postReq: {
    b: 11
  },
  c: [23, 25],
  extMsg: {
    d: 12
  }
};
```

Therefore we need to ensure that we remove field `a` from the string that is generated as a response of `renderURLSearchParams`

Reply:
cool. understand now. the confusion comes from `fieldList` as per comment later down the change

Reply:
Thank you for understanding 🙇🏼